### PR TITLE
Use macros to prevent redundant code in short lambdas.

### LINKED ISSUE #430: Way to avoid if/else
Instead of doing: 
`  if (c->GetDeviceType()  == kCpu) {  loop; } else {  declare lambda; eval }`
I propose to do something like this:
```
#define EVAL(context, dim, lambda_name, ...)  \
    if (context->GetDeviceType() == kCpu) {
      auto lambda_name = [=] __VA_ARGS__;
      for (int32_t i = 0; i < dim; i++) lambda_name(i);
  } else {
      auto lambda_name = [=] __device__ __VA_ARGS__;
     EvalDevice(context, dim, lambda_name);
 }
```
which would be called as:
```
  EVAL(c, foo.dim, lambda_do_something, (int32_t i) -> void { code here; });
```

### PR DESCRIPTION
**Pull request by csukuangfj**
Closes #430 

@danpovey 
Do we need to pass `lambda_name` as a macro aguement? I think it can be omitted and we can generate an unique name
inside the lambda.

### CODE REVIEW

**Comment by danpovey** [comment]
The reason to pass the lambda_name is for better debuggability.

File: k2/csrc/macros.h [hunk 1]
```diff
@@ -21,4 +21,16 @@
 #define K2_FUNC __func__
 #endif
 
+#define K2_EVAL(context, dim, lambda_name, arg_name, ...)              \
```
Comment (danpovey) [inline k2/csrc/macros.h:24]: Need some documentation for this macro, please!

Comment (danpovey) [reply k2/csrc/macros.h:24]: ... also please try using it at least once.
Eventually we'll have a K2_EVAL2 macro, for 2-arg lambdas.

Comment (csukuangfj) [reply k2/csrc/macros.h]: Documentation is added.

`K2_EVAL2` is also added.

I've replaced nearly all ocurrences of Eval and Eval2 with K2_EVAL and K2_EVAL2.

**Comment by danpovey** [comment]
Thanks!  Let me wait to merge this, till I merge the intersect code, or there will be bad merge problems.

**Comment by csukuangfj** [comment]
Ok, I'll resolve any conflicts introduced by `intersect`.

**Comment by csukuangfj** [comment]
I'll review all the changes and remove WIP tomorrow morning.

**Comment by danpovey** [comment]
Thanks!

File: k2/csrc/intersect_pruned.cu [hunk 2]
```diff
@@ -730,81 +735,84 @@ class MultiGraphDenseIntersectPruned {
...
     }
     {
       NVTX_RANGE("LambdaResetStateMap");
       const int32_t *next_states_row_ids1 = ans->states.shape.RowIds(1).Data();
-      auto lambda_reset_state_map =
-          [=] __host__ __device__(int32_t state_idx01) -> void {
-        int32_t a_fsas_state_idx01 =
-           kept_states_data[state_idx01].a_fsas_state_idx01,
-           fsa_idx0 = next_states_row_ids1[state_idx01];
-        K2_CHECK_EQ(state_map_acc(fsa_idx0, a_fsas_state_idx01), state_idx01);
```
Comment (csukuangfj) [inline k2/csrc/intersect_pruned.cu:803]: I changed this CHECK to **D**CHCECK as it is in a lambda, i.e., inside a `for` loop.

File: k2/csrc/intersect_pruned.cu [hunk 3]
```diff
@@ -879,48 +887,47 @@ class MultiGraphDenseIntersectPruned {
...
+                    oshape_arc_idx01x = oshape_row_splits2[oshape_arc_idx01],
+                    oshape_arc_idx01xx = oshape_row_splits3[oshape_arc_idx01x],
+                    oshape_arc_idx0123 = oshape_arc_idx01xx + arcs_idx12;
+            // note, for the previous line: indexes 1 and 2 of FrameInfo::arcs
+            // (==state,arc) become indexes 2 and 3 of oshape_unpruned_.
+            keep_arcs_data[oshape_arc_idx0123] = keep_this_arc;
+            arc_backward_prob_data[arcs_idx012] = backward_loglike;
+          });
     } else {
-      assert(arc_backward_prob.NumElements() == 0 &&
```
Comment (csukuangfj) [inline k2/csrc/intersect_pruned.cu:921]: All `assert` statements are changed to ``K2_DCHECK_*`.

File: k2/csrc/top_sort.cu [hunk 4]
```diff
@@ -286,30 +286,29 @@ class TopSorter {
...
         RaggedShape2(&has_final_state, nullptr, n), Array1<int32_t>(c_, n));
     int32_t *ans_data = ans->values.Data();
     const int32_t *ans_row_ids1_data = ans->RowIds(1).Data();
-    auto lambda_set_final_state = [=] __host__ __device__(int32_t i) -> void {
-      int32_t fsa_idx0 = ans_row_ids1_data[i],
-              final_state = fsas_row_splits1_data[fsa_idx0 + 1] - 1;
-      // If the following fails, it likely means an input FSA was invalid (e.g.
-      // had exactly one state, which is not allowed).  Either that, or a code
-      // error.
-      K2_CHECK_GT(final_state, fsas_row_splits1_data[fsa_idx0]);
```
Comment (csukuangfj) [inline k2/csrc/top_sort.cu:309]: K2_CHECK_GT is changed to K2_DCHECK_GT.

**Comment by csukuangfj** [comment]
Ready for review and merge.
It contains lots of `diff`, but most of them are from replacing `Eval` and `Eval2` with `K2_EVAL` and `K2_EVAL2`.

**Pull request merged**